Fix predicate failure for negative offset value on _timestamp column

harris233 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->

<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?
A fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
Kafka connector

> How would you describe this change to a non-technical end user or system administrator?
Fix predicate failure for negative offset value on _timestamp column

## Related issues, pull requests, and links

Fixes #13167

## Documentation

(x) No documentation is needed.

## Release notes

(x) Release notes entries required with the following suggested text:

Issue:[#13167](https://github.com/trinodb/trino/issues/13167)

ebyhr reviewed (commented):
Thanks for you contribution! Please submit CLA if you haven't sent it https://github.com/trinodb/cla.

Could you add a test to `TestKafkaIntegrationPushDown`?

Praveen2112 commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:
```diff
@@ -129,11 +129,15 @@ public KafkaFilteringResult getKafkaFilterResult(
                     Optional<Range> finalOffsetTimestampRanged = offsetTimestampRanged;
-                    partitionBeginOffsets = overridePartitionBeginOffsets(partitionBeginOffsets,
-                            partition -> findOffsetsForTimestampGreaterOrEqual(kafkaConsumer, partition, finalOffsetTimestampRanged.get().getBegin()));
+                    if (offsetTimestampRanged.get().getBegin() >= 0) {
```
Can we add some tests to check if it works as expected ?

harris233 replied:
OK, I added some test cases in TestKafkaIntegrationPushDown.

harris233 commented:
> Thanks for you contribution! Please submit CLA if you haven't sent it https://github.com/trinodb/cla.
> 
> Could you please add a test to `TestKafkaIntegrationPushDown`? Also, I would recommend improving the commit title - we can't understand what's the issue from "small bug".
OK, I sent cla to [cla@trino.io](mailto:cla@trino.io) just now. I have added some test cases and updated the commit title.

harris233 commented:
Hello, I have signed CLA and got it.Please review again@[ebyhr](https://github.com/ebyhr)

ebyhr commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:
```diff
@@ -129,11 +129,15 @@ public KafkaFilteringResult getKafkaFilterResult(
                     Optional<Range> finalOffsetTimestampRanged = offsetTimestampRanged;
-                    partitionBeginOffsets = overridePartitionBeginOffsets(partitionBeginOffsets,
-                            partition -> findOffsetsForTimestampGreaterOrEqual(kafkaConsumer, partition, finalOffsetTimestampRanged.get().getBegin()));
+                    if (offsetTimestampRanged.get().getBegin() >= 0) {
```
It would better to extract a constant for 0. e.g. `MIN_ALLOWED_OFFSET`

harris233 replied:
OK, I used `INVALID_KAFKA_RANGE_INDEX`

ebyhr commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationPushDown.java:
```diff
@@ -169,6 +169,57 @@ public void testTimestampLogAppendModePushDown()
     }
 
+    @Test
+    public void testTimestampCreateTimeModeOpenRangePushDown()
```
Please add missing operators `=`, `>`, `<=` and `BETWEEN`. The above change affects them.

harris233 replied:
OK, I added these cases in test.

ebyhr reviewed (commented):
Could you fix the commit title likes below sentence?  
```
Fix predicate failure for negative offset value on _timestamp column
```

harris233 commented:
> Could you fix the commit title likes below sentence?
> 
> ```
> Fix predicate failure for negative offset value on _timestamp column
> ```

OK, I changed to `Fix predicate failure for negative offset value on _timestamp column`

ebyhr commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:
```diff
@@ -129,11 +129,15 @@ public KafkaFilteringResult getKafkaFilterResult(
                     Optional<Range> finalOffsetTimestampRanged = offsetTimestampRanged;
-                    partitionBeginOffsets = overridePartitionBeginOffsets(partitionBeginOffsets,
-                            partition -> findOffsetsForTimestampGreaterOrEqual(kafkaConsumer, partition, finalOffsetTimestampRanged.get().getBegin()));
+                    if (offsetTimestampRanged.get().getBegin() > INVALID_KAFKA_RANGE_INDEX) {
```
It would be better to leave comment why we ignore instead of an exception.

harris233 replied:
OK, I added the reason why we need.

harris233 commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationPushDown.java:
```diff
@@ -115,26 +114,9 @@ private void assertProcessedInputPossitions(String sql, long expectedProcessedIn
-    @Test
-    public void testTimestampCreateTimeModePushDown()
-            throws Exception
+    private void assertProcessedInputPossitionsByUsingTimestampUpperPushDown(String sql, long expectedProcessedInputPositions)
```
OK. I added a session argument in assertProcessedInputPositions.

ebyhr commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationPushDown.java:
```diff
@@ -101,70 +100,60 @@ public void testPartitionPushDown()
-            MaterializedResultWithQueryId queryResult = queryRunner.executeWithQueryId(getSession(), sql);
-            assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), expectedProcessedInputPositions);
-        });
+        Session session = getSession();
```
We don't need to pass the default session by adding another `assertProcessedInputPositions`. See other comment.

harris233 replied:
OK

ebyhr commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationPushDown.java:
```diff
@@ -101,70 +100,60 @@ public void testPartitionPushDown()
     public void testTimestampCreateTimeModePushDown()
             throws Exception
     {
+        Session session = getSession();
```
ditto.

harris233 replied:
OK

ebyhr commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationPushDown.java:
```diff
@@ -101,70 +100,60 @@ public void testPartitionPushDown()
-            assertThat(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions())
-                    .isEqualTo(2);
-        });
+        session = Session.builder(getSession())
```
Let's rename to `sessionWithUpperBoundPushDownEnabled`

harris233 replied:
I used sessionWithUpperBoundPushDownEnabled instead of session

ebyhr commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationPushDown.java:
```diff
@@ -101,70 +100,60 @@ public void testPartitionPushDown()
+        assertProcessedInputPossitions(format("SELECT count(*) FROM default.%s WHERE _timestamp = timestamp '%s'", topicNameLogAppend, recordMessage.getStartTime()), 0, session);
+    }
 
+    private void assertProcessedInputPossitions(String sql, long expectedProcessedInputPositions, Session session)
```
Let's add a preparatory commit to fix typo. `assertProcessedInputPositions`

harris233 replied:
I changed to assertProcessedInputPositions

ebyhr replied:
> Let's add a preparatory commit to fix typo.

This commend isn't addressed.

harris233 replied:
I added another commit to fix it.

ebyhr commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationPushDown.java:
```diff
@@ -101,70 +100,60 @@ public void testPartitionPushDown()
-            assertThat(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions())
-                    .isEqualTo(2);
+            MaterializedResultWithQueryId queryResult = queryRunner.executeWithQueryId(session, sql);
+            assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), expectedProcessedInputPositions);
```
Let's add another `assertProcessedInputPositions` so that we can avoid passing the default session. 

```java
    private void assertProcessedInputPositions(String sql, long expectedProcessedInputPositions)
    {
        assertProcessedInputPositions(sql, expectedProcessedInputPositions, getSession());
    }

    private void assertProcessedInputPositions(String sql, long expectedProcessedInputPositions, Session session)
    {
        DistributedQueryRunner queryRunner = getDistributedQueryRunner();
        assertEventually(() -> {
            MaterializedResultWithQueryId queryResult = queryRunner.executeWithQueryId(session, sql);
            assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), expectedProcessedInputPositions);
        });
    }
```

harris233 replied:
OK, I added it.

ebyhr commented:
@harris233 Please rebase instead of merge to follow upstream.

harris233 commented:
> @harris233 Please rebase instead of merge to follow upstream.

OK, I rebased it.

ebyhr commented:
Merged, thanks!
